plugins/module_utils/mh/mixins/dest.py

```diff
@@ -0,0 +1,356 @@
+
+    def __process_data__(self):
+        # type: () -> None
+        """
+        Implement this in your module to define how to process data the
+        destination file.
+
+        When you run this method, you get binary raw data from your destination
+        file into `self.raw_content` method.
+
+        After you process it, do not write the result directly in the
+        destination file (it is automatically done later with the
+        content of `self.raw_content`) instead, put your result in binary form
+        `self.raw_content`.
+
+        It is expected that this method manage the module change flag if data
+        was updated. A good way do de this is by adding an item in `self.vars`
+        which can handle diffs and changes on data.
+        """
+        raise NotImplementedError()
+
+    def __quit_module__(self):
+        # type: () -> None
+        self.__update_dest__()
+
+    def __update_dest__(self):
+        # type: () -> None
+        if self.has_changed():
+            self._write_temp_file()
+            self._backup_dest()
+            self._replace_dest_with_temp()
+            self.vars.set('created', self._dest_infos['can_create'])
+        self._update_dest_fs_attributes()
+
+    @check_mode_skip
+    def _write_temp_file(self):
+        # type: () -> None
+        """
+        Write the binary data from `self.raw_content` into an Ansible temp
+        file.
+        """
+        fd, self._tmpfile = tempfile.mkstemp(dir=self.module.tmpdir)
+        os.write(fd, self.raw_content)
+        self.module.add_cleanup_file(self._tmpfile)
+        os.close(fd)
+
+    @check_mode_skip
+    def _backup_dest(self):
+        # type: () -> None
+        if self._dest_infos['can_backup'] and self.vars.get('backup'):
+            self.vars.set('backup_file',
+                          self.module.backup_local(self.vars['path']))
+
+    @cause_changes(on_success=True)
+    @check_mode_skip
+    def _replace_dest_with_temp(self):
+        # type: () -> None
+        try:
+            self.module.atomic_move(self._tmpfile, self.vars['path'])
+        except Exception as ex:
+            if self.vars.get('backup_file'):
+                self.module.add_cleanup_file(self.vars.backup_file)
+            raise ex
+
+    def _update_dest_fs_attributes(self):
+        # type: () -> None
+        attr_changes = {'before': {}, 'after': {}}
+        self.module.set_fs_attributes_if_different(self._dest_attrs,
+                                                   self.changed,
+                                                   diff=attr_changes)
+        for key, val in attr_changes['after'].items():
+            if key == 'dest' or key == 'path':
+                continue
+            if val is not None:
+                self.vars.meta(key).set(diff=True, change=True, output=True)
+                self.vars.set(key, val)
+        self.changed = self.has_changed()
+
+    @classmethod
+    def with_default_params(cls, params):
```

Comment:
@felixfontein 
> Out of curiosity: why not use `add_file_common_args=true` (https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/basic.py#L438)

It is used indirectly. When you create a new module by using `DestFileModuleHelper`, you can use use the `with_default_params` method to add a set of default parameters that add `add_file_common_args=True`. You can see an implementation of this in `MDestFile` (https://github.com/ansible-collections/community.general/pull/4918/files#diff-5c9198faa2254aed31f5a0df42c57d02f694b5d72add25b2b711416bd217768d).